op-e2e/actions/span_batch_test.go

```diff
@@ -202,6 +312,103 @@ func TestAcceptSingularBatchAfterHardfork(gt *testing.T) {
 	require.NotNil(t, vTx)
 }
 
+// TestMixOfBatchesAfterHardfork tests behavior of op-node after Delta hardfork.
+// op-node must accept SingularBatch and SpanBatch in sequence.
+func TestMixOfBatchesAfterHardfork(gt *testing.T) {
+	t := NewDefaultTesting(gt)
+	p := &e2eutils.TestParams{
+		MaxSequencerDrift:   20, // larger than L1 block time we simulate in this test (12)
+		SequencerWindowSize: 24,
+		ChannelTimeout:      20,
+		L1BlockTime:         12,
+	}
+	minTs := hexutil.Uint64(0)
+	dp := e2eutils.MakeDeployParams(t, p)
+
+	// Activate Delta hardfork for verifier.
+	dp.DeployConfig.L2GenesisDeltaTimeOffset = &minTs
+	sd := e2eutils.Setup(t, dp, defaultAlloc)
+	log := testlog.Logger(t, log.LvlError)
+	miner, seqEngine, sequencer := setupSequencerTest(t, sd, log)
+	verifEngine, verifier := setupVerifier(t, sd, log, miner.L1Client(t, sd.RollupCfg), &sync.Config{})
+	rollupSeqCl := sequencer.RollupClient()
+	seqEngCl := seqEngine.EthClient()
+
+	// Deploy config for span batch disabled batcher
+	dp2 := e2eutils.MakeDeployParams(t, p)
+	dp2.DeployConfig.L2GenesisDeltaTimeOffset = nil
+	sd2 := e2eutils.Setup(t, dp2, defaultAlloc)
+
+	sequencer.ActL2PipelineFull(t)
+	verifier.ActL2PipelineFull(t)
+	miner.ActEmptyBlock(t)
+
+	var txHashes [4]common.Hash
+	for i := 0; i < 4; i++ {
+		// Alice makes a L2 tx
+		n, err := seqEngCl.PendingNonceAt(t.Ctx(), dp.Addresses.Alice)
+		require.NoError(t, err)
+		signer := types.LatestSigner(sd.L2Cfg.Config)
+		tx := types.MustSignNewTx(dp.Secrets.Alice, signer, &types.DynamicFeeTx{
+			ChainID:   sd.L2Cfg.Config.ChainID,
+			Nonce:     n,
+			GasTipCap: big.NewInt(2 * params.GWei),
+			GasFeeCap: new(big.Int).Add(miner.l1Chain.CurrentBlock().BaseFee, big.NewInt(2*params.GWei)),
+			Gas:       params.TxGas,
+			To:        &dp.Addresses.Bob,
+			Value:     e2eutils.Ether(2),
+		})
+		require.NoError(gt, seqEngCl.SendTransaction(t.Ctx(), tx))
+		txHashes[i] = tx.Hash()
+
+		// Make L2 block
+		sequencer.ActL1HeadSignal(t)
+		sequencer.ActL2StartBlock(t)
+		seqEngine.ActL2IncludeTx(dp.Addresses.Alice)(t)
+		sequencer.ActL2EndBlock(t)
+		sequencer.ActBuildToL1Head(t)
+
+		// Select batcher mode
+		rcfg := sd.RollupCfg // SpanBatch mode
+		if i&2 == 1 {
+			rcfg = sd2.RollupCfg // SingularBatch mode
+		}
```

Comment:
Have a toggle in the batcher actor to force it to use a particular span batch mode may simplify this too (though I could wrong on that...)